app/models/manageiq/providers/kubevirt/infra_manager/provision/cloning.rb

```diff
@@ -80,58 +80,45 @@ def user_options(options)
   end
 
   def values(template, options)
-    default_params = {}
-    template.parameters.each do |param|
-      name = param[:name].downcase
+    template.parameters.each_with_object({}) do |param, default_params|
+      name  = param[:name].downcase
       value = options[name.to_sym]
-      if value && name == "memory"
-        value = "#{value}Mi"
-      end
+      value = "#{value}Mi" if value && name == "memory"
 
       default_params[name] = value || param[:value]
     end
-    default_params
   end
 
   def param_substitution!(object, params)
-    result = object
-    case result
+    case object
     when Hash
-      result.each do |k, v|
-        result[k] = param_substitution!(v, params)
-      end
+      object.transform_values! { |v| param_substitution!(v, params) }
     when Array
-      result.map { |v| param_substitution!(v, params) }
+      object.map! { |v| param_substitution!(v, params) }
     when String
-      result = sub_specific_object(params, object)
+      substitute_string_param!(object, params)
     end
-    result
   end
 
   #
   # Performs substitution on specific object.
   #
-  # @params params [Hash] Containing parameter names and values used for substitution.
   # @params object [String] Object on which substitution takes place.
+  # @params params [Hash] Containing parameter names and values used for substitution.
   # @returns [String] The outcome of substitution.
   #
-  def sub_specific_object(params, object)
-    result = object
+  def substitute_string_param!(object, params)
```

Comment:
I don't think substitute _string_ param is a good name though since this does more than just strings

Reply:
ah ok - glad you dug in - from the original code it felt like it was only called in the string case in the case statement, so that's why I suggested string in the name.  Up to you what to call it.

Reply:
Yeah the String case is when the value from the object that we're substituting is a string like: `:spec => {:domain => "${CPU}"}` it is the "${CPU}" case.  Then in substitute_string_param it is looking at what is being substituted.